### PR TITLE
Feature: 공지사항 페이지에 필요한 기능들 구현

### LINKED ISSUE #6: 💡 Feature: 서버 시간 조회 기능 구현
## Key Feature

현재 서버 시간을 반환하는 기능을 구현한다.

### Feature List

- [x] 현재 서버 시간 반환

### LINKED ISSUE #7: 💡 Feature: 학생 정보 반환 기능 구현
## Key Feature

현재 로그인한 학생 정보를 반환하는 기능을 구현한다.

### Feature List

- [x] 인터셉터는 클라이언트의 쿠키의 key 가 세션에 존재하는지를 체크한다.
- [x] argument resolver 에서 클라이언트의 쿠키를 key 로 하여 세션의 student의 PK 를 얻어 반환한다.
- [x] 학생의 이름, 학번, 소속, 학년을 반환하는 기능을 구현한다.

### LINKED ISSUE #8: 💡 Feature: 세션 연장 기능 구현
## Key Feature

서버에서의 세션 접속 시간을 연장하는 기능을 구현한다.

### Feature List

- [x] 서버에서의 세션 접속 시간을 연장하는 기능을 구현한다.

### PR DESCRIPTION
**Pull request by ghkdgus29**
## Key Feature

- 현재 서버 시간을 조회하는 기능을 구현하였습니다.
- 현재 세션의 지속 시간을 갱신하는 기능을 구현하였습니다.
- 학생 정보를 반환하는 기능을 구현하였습니다.
- 로그아웃 기능을 구현하였습니다.

## To reviewer

공지사항 기능이 크게 복잡하지 않아 금방 구현했습니다~

로그인을 체크하는 인터셉터에서는 클라이언트가 보낸 세션 쿠키의 value 가, 우리 서버의 세션의 key 로서 존재하는지 여부를 체크한 뒤 매핑되는 세션의 학생 엔티티의 PK 값을 request 에 넣어줍니다.

이후 argument resolver 를 사용하여 컨트롤러 단에서 studentPk 값 만 꺼내서 사용할 수 있습니다.

### CODE REVIEW

**Comment by ghkdgus29** [comment]
### Test Coverage Report
|Overall Project|95.13% **`-1.77%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|96.37%|:green_apple:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[StudentInformation.java](https://github.com/mujik-tigers/course-registration-system/blob/7eecb22022abba9341da7b1510d64795e5c25271/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fstudent%2Fdto%2FStudentInformation.java)|100%|:green_apple:|
|[AuthController.java](https://github.com/mujik-tigers/course-registration-system/blob/7eecb22022abba9341da7b1510d64795e5c25271/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fauth%2Fpresentation%2FAuthController.java)|100%|:green_apple:|
|[Grade.java](https://github.com/mujik-tigers/course-registration-system/blob/7eecb22022abba9341da7b1510d64795e5c25271/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fstudent%2FGrade.java)|100%|:green_apple:|
|[CurrentServerTime.java](https://github.com/mujik-tigers/course-registration-system/blob/7eecb22022abba9341da7b1510d64795e5c25271/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fclock%2Fdto%2FCurrentServerTime.java)|100%|:green_apple:|
|[ClockController.java](https://github.com/mujik-tigers/course-registration-system/blob/7eecb22022abba9341da7b1510d64795e5c25271/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fclock%2Fpresentation%2FClockController.java)|100%|:green_apple:|
|[StudentService.java](https://github.com/mujik-tigers/course-registration-system/blob/7eecb22022abba9341da7b1510d64795e5c25271/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fstudent%2Fapplication%2FStudentService.java)|100%|:green_apple:|
|[ResponseMessage.java](https://github.com/mujik-tigers/course-registration-system/blob/7eecb22022abba9341da7b1510d64795e5c25271/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Futil%2Fapi%2FResponseMessage.java)|100%|:green_apple:|
|[NonexistenceStudentException.java](https://github.com/mujik-tigers/course-registration-system/blob/7eecb22022abba9341da7b1510d64795e5c25271/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fexception%2Fstudent%2FNonexistenceStudentException.java)|100%|:green_apple:|
|[WebConfig.java](https://github.com/mujik-tigers/course-registration-system/blob/7eecb22022abba9341da7b1510d64795e5c25271/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fconfig%2FWebConfig.java)|100%|:green_apple:|
|[ErrorType.java](https://github.com/mujik-tigers/course-registration-system/blob/7eecb22022abba9341da7b1510d64795e5c25271/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fexception%2FErrorType.java)|100%|:green_apple:|
|[StudentController.java](https://github.com/mujik-tigers/course-registration-system/blob/7eecb22022abba9341da7b1510d64795e5c25271/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fstudent%2Fpresentation%2FStudentController.java)|100%|:green_apple:|
|[NonexistenceSessionCookieException.java](https://github.com/mujik-tigers/course-registration-system/blob/7eecb22022abba9341da7b1510d64795e5c25271/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fexception%2Fauth%2FNonexistenceSessionCookieException.java)|100%|:green_apple:|
|[LoginCheckInterceptor.java](https://github.com/mujik-tigers/course-registration-system/blob/7eecb22022abba9341da7b1510d64795e5c25271/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Finterceptor%2FLoginCheckInterceptor.java)|90.32% **`-9.68%`**|:green_apple:|
|[Student.java](https://github.com/mujik-tigers/course-registration-system/blob/7eecb22022abba9341da7b1510d64795e5c25271/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fstudent%2FStudent.java)|88.68% **`-11.32%`**|:green_apple:|
|[LoginArgumentResolver.java](https://github.com/mujik-tigers/course-registration-system/blob/7eecb22022abba9341da7b1510d64795e5c25271/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Futil%2Fresolver%2FLoginArgumentResolver.java)|86.84% **`-13.16%`**|:green_apple:|
|[Department.java](https://github.com/mujik-tigers/course-registration-system/blob/7eecb22022abba9341da7b1510d64795e5c25271/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fdepartment%2FDepartment.java)|80%|:green_apple:|
|[ProjectConstant.java](https://github.com/mujik-tigers/course-registration-system/blob/7eecb22022abba9341da7b1510d64795e5c25271/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Futil%2FProjectConstant.java)|0%|:x:|

File: src/main/java/site/courseregistrationsystem/clock/presentation/ClockController.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import lombok.RequiredArgsConstructor;
+import site.courseregistrationsystem.clock.dto.CurrentServerTime;
+import site.courseregistrationsystem.util.api.ApiResponse;
+import site.courseregistrationsystem.util.api.ResponseMessage;
+
+@RestController
+@RequiredArgsConstructor
```
Comment (yeonise) [inline src/main/java/site/courseregistrationsystem/clock/presentation/ClockController.java]: 💁🏻 현재 생성자 주입을 받는 필드가 없어서 지워도 될 것 같아요!

File: src/test/java/site/courseregistrationsystem/RestDocsSupport.java [hunk 2]
```diff
@@ -1,54 +1,77 @@
...
+		given(cookieProperties.getDomain()).willReturn(cookieDomain);
+		given(cookieProperties.getPath()).willReturn(cookiePath);
+		given(cookieProperties.getExpiry()).willReturn(cookieExpiry);
+
+		given(sessionManager.fetch(any(String.class)))
+			.will(invocation -> {
+				String sessionId = invocation.getArgument(0, String.class);
+
+				if (sessionId.length() == 36) {
+					return new StudentSession(sessionId, 1L, 1000L);
```
Comment (yeonise) [inline src/test/java/site/courseregistrationsystem/RestDocsSupport.java]: 💁🏻 세션 지속 시간도 `3600L` 로 수정해도 될까요?
세션 테스트에 시간 검증을 추가하고 싶어서요

File: src/main/java/site/courseregistrationsystem/auth/presentation/AuthController.java [hunk 3]
```diff
@@ -22,22 +27,56 @@ public class AuthController {
...
+		response.addCookie(cookie);
+
+		return ApiResponse.of(HttpStatus.CREATED, ResponseMessage.LOGIN_SUCCESS.getMessage(), null);
+	}
 
-		Cookie cookie = new Cookie(cookieProperties.getName(), session.getId());  // 쿠키 생성
+	@ResponseStatus(HttpStatus.CREATED)
+	@PostMapping("/session")
+	public ApiResponse<Void> renewSessionDuration(@CookieValue(value = SESSION_ID) Cookie sessionCookie, HttpServletResponse response) {
+		StudentSession renewSession = sessionManager.renew(sessionCookie.getValue());
```
Comment (yeonise) [inline src/main/java/site/courseregistrationsystem/auth/presentation/AuthController.java:46]: 사실 시간을 갱신하는 것뿐만 아니라 세션 자체를 새로 만들긴합니다 💁🏻

File: src/main/java/site/courseregistrationsystem/auth/presentation/AuthController.java [hunk 4]
```diff
@@ -22,22 +27,56 @@ public class AuthController {
...
+	}
 
-		response.addCookie(cookie); // 쿠키 설정
-
-		return ApiResponse.of(HttpStatus.CREATED, ResponseMessage.LOGIN_SUCCESS.getMessage(), null);
+	private void invalidateCookie(Cookie sessionCookie) {
+		sessionCookie.setHttpOnly(true);
+		sessionCookie.setSecure(true);
+		sessionCookie.setDomain(cookieProperties.getDomain());
+		sessionCookie.setPath(cookieProperties.getPath());
```
Comment (yeonise) [inline src/main/java/site/courseregistrationsystem/auth/presentation/AuthController.java:78]: 💁🏻 이 부분은 다시 설정하지 않아도 될 것 같아요!

File: src/main/java/site/courseregistrationsystem/util/ProjectConstant.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package site.courseregistrationsystem.util;
+
+public abstract class ProjectConstant {
+
+	public static final String STUDENT_PK = "studentPK";
```
Comment (yeonise) [inline src/main/java/site/courseregistrationsystem/util/ProjectConstant.java]: 💁🏻 다른 곳에서 사용한 `studentPk` 나 `STUDENTPK`로 통일해도 될까요?

File: src/main/java/site/courseregistrationsystem/util/ProjectConstant.java [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package site.courseregistrationsystem.util;
+
+public abstract class ProjectConstant {
```
Comment (yeonise) [inline src/main/java/site/courseregistrationsystem/util/ProjectConstant.java:3]: 💁🏻 프로젝트 내 모든 상수를 관리하는 클래스인가요? 아니면 Session or Cookie 관련 상수만 관리하는 건가요?

File: src/main/java/site/courseregistrationsystem/department/infrastructure/DepartmentRepository.java [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package site.courseregistrationsystem.department.infrastructure;
+
+import org.springframework.data.jpa.repository.JpaRepository;
+
+import site.courseregistrationsystem.department.Department;
+
+public interface DepartmentRepository extends JpaRepository<Department, Long> {
+
+}
```
Comment (yeonise) [inline src/main/java/site/courseregistrationsystem/department/infrastructure/DepartmentRepository.java:9]: 테스트 코드를 위한 레포지토리 생성은 고민해보았으면 좋겠다는 피드백을 받은 적이 있어서 더 좋은 방법이 있는지 고민해볼게요 💁🏻

**Review by yeonise — APPROVED** [review]
빠르게 여러 기능을 완료하셨네요 cookie 검증까지 👏🏻 고생하셨습니다! 머지해도 될 것 같아요

File: src/main/java/site/courseregistrationsystem/clock/presentation/ClockController.java (hunk 1 above)
Comment (ghkdgus29) [reply src/main/java/site/courseregistrationsystem/clock/presentation/ClockController.java]: ok~

File: src/main/java/site/courseregistrationsystem/department/infrastructure/DepartmentRepository.java (hunk 7 above)
Comment (ghkdgus29) [reply src/main/java/site/courseregistrationsystem/department/infrastructure/DepartmentRepository.java:9]: em.createQuery 로 직접 쿼리문을 날려주는 방법도 생각은 해보았어요
근데 그러면 쿼리를 하드코딩해야하는데 그게 나을까요?

File: src/main/java/site/courseregistrationsystem/util/ProjectConstant.java (hunk 6 above)
Comment (ghkdgus29) [reply src/main/java/site/courseregistrationsystem/util/ProjectConstant.java:3]: 프로젝트 내 모든 상수를 관리하는 클래스에요~

File: src/main/java/site/courseregistrationsystem/util/ProjectConstant.java (hunk 5 above)
Comment (ghkdgus29) [reply src/main/java/site/courseregistrationsystem/util/ProjectConstant.java]: studentPk 로 바꿨습니다~

File: src/test/java/site/courseregistrationsystem/RestDocsSupport.java (hunk 2 above)
Comment (ghkdgus29) [reply src/test/java/site/courseregistrationsystem/RestDocsSupport.java]: 3600L 로 수정했습니다~

**Comment by ghkdgus29** [comment]
### Test Coverage Report
|Overall Project|95.13% **`-1.77%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|96.37%|:green_apple:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[StudentInformation.java](https://github.com/mujik-tigers/course-registration-system/blob/9d1670f8afe80ea77ba3563835558419185bdd27/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fstudent%2Fdto%2FStudentInformation.java)|100%|:green_apple:|
|[AuthController.java](https://github.com/mujik-tigers/course-registration-system/blob/9d1670f8afe80ea77ba3563835558419185bdd27/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fauth%2Fpresentation%2FAuthController.java)|100%|:green_apple:|
|[Grade.java](https://github.com/mujik-tigers/course-registration-system/blob/9d1670f8afe80ea77ba3563835558419185bdd27/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fstudent%2FGrade.java)|100%|:green_apple:|
|[CurrentServerTime.java](https://github.com/mujik-tigers/course-registration-system/blob/9d1670f8afe80ea77ba3563835558419185bdd27/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fclock%2Fdto%2FCurrentServerTime.java)|100%|:green_apple:|
|[ClockController.java](https://github.com/mujik-tigers/course-registration-system/blob/9d1670f8afe80ea77ba3563835558419185bdd27/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fclock%2Fpresentation%2FClockController.java)|100%|:green_apple:|
|[StudentService.java](https://github.com/mujik-tigers/course-registration-system/blob/9d1670f8afe80ea77ba3563835558419185bdd27/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fstudent%2Fapplication%2FStudentService.java)|100%|:green_apple:|
|[ResponseMessage.java](https://github.com/mujik-tigers/course-registration-system/blob/9d1670f8afe80ea77ba3563835558419185bdd27/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Futil%2Fapi%2FResponseMessage.java)|100%|:green_apple:|
|[NonexistenceStudentException.java](https://github.com/mujik-tigers/course-registration-system/blob/9d1670f8afe80ea77ba3563835558419185bdd27/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fexception%2Fstudent%2FNonexistenceStudentException.java)|100%|:green_apple:|
|[WebConfig.java](https://github.com/mujik-tigers/course-registration-system/blob/9d1670f8afe80ea77ba3563835558419185bdd27/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fconfig%2FWebConfig.java)|100%|:green_apple:|
|[ErrorType.java](https://github.com/mujik-tigers/course-registration-system/blob/9d1670f8afe80ea77ba3563835558419185bdd27/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fexception%2FErrorType.java)|100%|:green_apple:|
|[StudentController.java](https://github.com/mujik-tigers/course-registration-system/blob/9d1670f8afe80ea77ba3563835558419185bdd27/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fstudent%2Fpresentation%2FStudentController.java)|100%|:green_apple:|
|[NonexistenceSessionCookieException.java](https://github.com/mujik-tigers/course-registration-system/blob/9d1670f8afe80ea77ba3563835558419185bdd27/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fexception%2Fauth%2FNonexistenceSessionCookieException.java)|100%|:green_apple:|
|[LoginCheckInterceptor.java](https://github.com/mujik-tigers/course-registration-system/blob/9d1670f8afe80ea77ba3563835558419185bdd27/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Finterceptor%2FLoginCheckInterceptor.java)|90.32% **`-9.68%`**|:green_apple:|
|[Student.java](https://github.com/mujik-tigers/course-registration-system/blob/9d1670f8afe80ea77ba3563835558419185bdd27/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fstudent%2FStudent.java)|88.68% **`-11.32%`**|:green_apple:|
|[LoginArgumentResolver.java](https://github.com/mujik-tigers/course-registration-system/blob/9d1670f8afe80ea77ba3563835558419185bdd27/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Futil%2Fresolver%2FLoginArgumentResolver.java)|86.84% **`-13.16%`**|:green_apple:|
|[Department.java](https://github.com/mujik-tigers/course-registration-system/blob/9d1670f8afe80ea77ba3563835558419185bdd27/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Fdepartment%2FDepartment.java)|80%|:green_apple:|
|[ProjectConstant.java](https://github.com/mujik-tigers/course-registration-system/blob/9d1670f8afe80ea77ba3563835558419185bdd27/src%2Fmain%2Fjava%2Fsite%2Fcourseregistrationsystem%2Futil%2FProjectConstant.java)|0%|:x:|

**Pull request merged**